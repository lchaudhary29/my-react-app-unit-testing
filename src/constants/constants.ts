export const isEmployee: boolean = true;
