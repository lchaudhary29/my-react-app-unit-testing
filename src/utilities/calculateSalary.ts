// src/calculateSalary.ts

export function calculateSalary(
  baseSalary: number,
  bonuses: number[] = [],
  deductions: number[] = []
): number {
  const totalBonuses = bonuses.reduce((acc, bonus) => acc + bonus, 0);
  const totalDeductions = deductions.reduce(
    (acc, deduction) => acc + deduction,
    0
  );
  return baseSalary + totalBonuses - totalDeductions;
}
