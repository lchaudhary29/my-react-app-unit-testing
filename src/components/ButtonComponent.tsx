// src/ButtonComponent.tsx

import React, { useState } from 'react';
import { calculateSalary } from '../utilities/calculateSalary';
import { isEmployee } from '../constants/constants';

const ButtonComponent: React.FC = () => {
    const [text, setText] = useState('Click me');

    const baseSalary = 50000;
    const bonuses = [2000, 1500];
    const deductions = [1000, 500];

    const salary = calculateSalary(baseSalary, bonuses, deductions);

    if (isEmployee) {
        console.log(`The employee's total salary is: $${salary}`);
    }

    const handleClick = () => {
        setText('Clicked');
    };

    return <button onClick={handleClick}>{text}</button>;
};

export default ButtonComponent;
