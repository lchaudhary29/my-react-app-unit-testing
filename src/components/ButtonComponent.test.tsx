// src/ButtonComponent.test.tsx

import React from 'react';
import { shallow } from 'enzyme';
import ButtonComponent from './ButtonComponent';
// Import the module that contains calculateSalary
import * as CalculateSalaryModule from '../utilities/calculateSalary';
import * as ConstantModule from '../constants/constants';

describe('<ButtonComponent />', () => {
    // Spy on calculateSalary before the tests run
    let calculateSalarySpy: jest.SpyInstance<number, [baseSalary: number, bonuses?: number[] | undefined, deductions?: number[] | undefined]>;

    beforeEach(() => {
        // Use jest.spyOn to spy on the calculateSalary function
        calculateSalarySpy = jest.spyOn(CalculateSalaryModule, 'calculateSalary').mockReturnValue(100000);
        // Optional: Mock the implementation if needed
        // calculateSalarySpy.mockImplementation(() => 100000);

        Object.defineProperty(ConstantModule, "isEmployee", { value: true });

    });

    afterEach(() => {
        // Ensure the spy is restored to its original state after each test
        calculateSalarySpy.mockRestore();
    });

    it('renders correctly, and changes text on click', () => {
        const wrapper = shallow(<ButtonComponent />);


        expect(wrapper.text()).toBe('Click me');

        wrapper.find('button').simulate('click');
        expect(wrapper.text()).toBe('Clicked');
        expect(wrapper).toMatchSnapshot();
    });

    it('test calculateSalary', () => {

        calculateSalarySpy.mockRestore();
        // Call the function to test the spy
        const result = CalculateSalaryModule.calculateSalary(50000, [2000, 1500], [1000, 500]);
        // If you mocked the implementation, assert the mocked return value
        expect(result).toBe(52000);
    });
});
