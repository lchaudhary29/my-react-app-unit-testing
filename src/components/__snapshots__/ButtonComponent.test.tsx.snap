// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonComponent /> renders correctly, and changes text on click 1`] = `ShallowWrapper {}`;
